### DISCUSSION TITLE
Why can only continuous memory be used for the ucp_put_nbx/ucp_get_nbx interface?

### DISCUSSION THREAD

**Discussion by super-train** [openucx/ucx#10315] [Q&A]
Hello everyone, I have a question now. Can anyone help me？

When I use ucp_put_nbx or ucp_get-nbx in ucx to perform unilateral operations with rdma, I find that these two interfaces in ucx only support contiguous memory,
Ucp_put_nbx : can only send continuous memory data from the local end;
Ucp_get_nbx : can only receive remote data into continuous local memory;
Why do both interfaces require that the local address must be continuous,

According to the IB specification of RDMA, unilateral operations do not simply require that the destination address be continuous

My current requirement is how to use the UCX interface to perform unilateral operations using RDMA, as the local addresses are discrete 4k pages

Why does the ucp_tag_send_nbx interface support data of type ucp_dt_iov_t * iov

Looking forward to everyone's answers, thank you!

**Comment by yosefe** [comment]
Hi, which version of UCX are you using? The lastest versions should support IOV for local buffers.

**Reply by super-train** [reply]
Hello mate,

First of all, I sincerely apologize for the delayed response due to some reasons. Thank you for your reply. I am using UCX version ucx-1.12.1, but today I just downloaded the latest version ucx-1.18.0 and checked the latest source code. Taking ucp_pub_nbx as an example, in the interface, the default data type is still contig, which is ucp_datatype_t datatype = ucp_dt_make_contig(1); Just like in the original version 1.12.1.

Moreover, in the actual interface ucp_rma_basic_progress_put, when it uses zero-copy, the processing code is as follows:

c
uct_iov_t iov;

md_index = ucp_ep_md_index(ep, lane);
/* TODO: leave last fragment for bcopy */
packed_len = ucs_min(req->send.length, rma_config->max_put_zcopy);
/* TODO: use ucp_dt_iov_copy_uct */
iov.buffer = (void *)req->send.buffer;
iov.length = packed_len;
iov.count = 1; // Hardcoded here
iov.memh = req->send.state.dt.dt.contig.memh->uct[md_index];

status = UCS_PROFILE_CALL(uct_ep_put_zcopy,
                          ucp_ep_get_fast_lane(ep, lane), &iov, 1,
                          req->send.rma.remote_addr,
                          rkey->cache.rma_rkey,
                          &req->send.state.uct_comp);
Here, iov.count = 1 is hardcoded again, just like in version 1.12.1, and does not support iovec-type data.

Please help me check if my analysis is correct. If there are any issues, please let me know. Thank you!
